content/program-testing.md

```diff
@@ -0,0 +1,488 @@
+
+```tsx
+describe("begin tests", async () => {
+
+	// tests go here
+	...
+
+})
+```
+
+Inside the `describe` section, each test is designated with `it`
+
+```tsx
+describe("begin tests", async () => {
+    // first Mocha test
+    it('first test', async () => {
+			...
+		})
+		// second Mocha test
+		it('second test', async () => {
+			...
+		})
+		...
+})
+```
+
+The Chai package is used to determine whether or not each test passes, it has an `expect` function that can easily compare values. You would use Chai to verify that whatever operation your test was supposed to execute returns the expected value or updated some data correctly.
+
+```tsx
+describe("begin tests", async () => {
+    // first Mocha test
+    it('first test', async () => {
+		// initialization code here to send the transaction
+		...
+		// fetch account info and deserialize
+        const acct_info = await connection.getAccountInfo(pda)
+        const acct = acct_struct.decode(acct_info.data)
+
+        // compare the value in the account to what you expect it to be
+        chai.expect(acct.num).to.equal(1)
+    })
+})
+```
+
+Running `npm run test` will execute all of the tests inside the `describe` block and return something like this indicating whether or not each one has passed or failed.
+
+```
+> scripts@1.0.0 test
+> mocha -r ts-node/register ./test/*.ts
+
+  ✔ first test (1308ms)
+  ✔ second test
+
+  2 passing (1s)
+```
+
+## Debugging
+
+When testing a program, you may run into errors or output that is unexpected, that’s where debugging comes into play. Debugging refers to the process of finding and resolving errors or unintended behavior of software programs.
+
+### Error Codes
+
+When writing programs in general, it is inevitable that you will spend a good portion of your time debugging. There are program errors that you may come across from time to time and they can be a little confusing. Some program errors have a hexadecimal code associated with them that looks something along the lines of `0xa4` which might not seem to make any sense. This error code is actually a hexadecimal representation of this error’s decimal index inside the error enum of the program that returned it. So, if an error is returned in your program with a hexadecimal code and no message, try converting that hexadecimal number *x* to decimal *y* and looking up the corresponding error at *y* index of the program’s error enum.
+
+### Program Logs
+
+Another helpful tool when debugging is logging. Solana makes it very easy to create new custom logs with the `msg!()` macro and, as you’ve seen in this course, you can even log data from accounts inside the program. This is probably one of the most helpful tools when it comes to debugging because you can see exactly how your program is interacting with the accounts involved and if it’s doing what is expected by logging data throughout the program.
```

Comment:
I found last night that I could use `println!` in the unit tests, which makes sense since it's not run on chain. It may be worth mentioning this as well.

Reply:
oh nice I'll try that out, good find!